src/resolve/resolved-dns-browse-services.c

```diff
@@ -0,0 +1,779 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "af-list.h"
+#include "event-util.h"
+#include "dns-domain.h"
+#include "json-util.h"
+#include "random-util.h"
+#include "resolved-dns-browse-services.h"
+#include "resolved-dns-cache.h"
+#include "resolved-dns-question.h"
+#include "resolved-dns-query.h"
+#include "resolved-dns-rr.h"
+#include "resolved-dns-scope.h"
+#include "resolved-manager.h"
+#include "resolved-varlink.h"
+#include "string-table.h"
+
+typedef enum BrowseServiceUpdateEvent {
+        BROWSE_SERVICE_UPDATE_ADDED,
+        BROWSE_SERVICE_UPDATE_REMOVED,
+        _BROWSE_SERVICE_UPDATE_MAX,
+        _BROWSE_SERVICE_UPDATE_INVALID = -EINVAL,
+} BrowseServiceUpdateEvent;
+
+static const char * const browse_service_update_event_table[_BROWSE_SERVICE_UPDATE_MAX] = {
+        [BROWSE_SERVICE_UPDATE_ADDED]   = "added",
+        [BROWSE_SERVICE_UPDATE_REMOVED] = "removed",
+};
+
+DEFINE_PRIVATE_STRING_TABLE_LOOKUP_TO_STRING(browse_service_update_event, BrowseServiceUpdateEvent);
+
+/* RFC6762 5.2
+ * The intervals between successive queries MUST increase by at least a
+ * factor of two. When the interval between queries reaches or exceeds
+ * 60 minutes, perform subsequent queries at a steady-state rate of one
+ * query per hour. */
+usec_t mdns_calculate_next_query_delay(usec_t current_delay) {
+        assert(current_delay <= 60 * 60 * USEC_PER_SEC);
+
+        if (current_delay == 0)
+                return USEC_PER_SEC;
+
+        return current_delay < 60 * 60 / 2 * USEC_PER_SEC ? current_delay * 2 : 60 * 60 * USEC_PER_SEC;
+}
+
+/* RFC 6762 section 5.2
+ * The querier should plan to issue a query at 80% of
+ * the record lifetime, and then if no answer is received, at 85%, 90%, and 95%. */
+static inline int DNS_RECORD_TTL_STATE_TO_PERCENT(DnsRecordTTLState ttl_state) {
+        switch (ttl_state) {
+        case DNS_RECORD_TTL_STATE_80_PERCENT:
+                return 80;
+        case DNS_RECORD_TTL_STATE_85_PERCENT:
+                return 85;
+        case DNS_RECORD_TTL_STATE_90_PERCENT:
+                return 90;
+        case DNS_RECORD_TTL_STATE_95_PERCENT:
+                return 95;
+        case DNS_RECORD_TTL_STATE_100_PERCENT:
+                return 100;
+        default:
+                return -1;
+        }
+}
+
+usec_t mdns_maintenance_next_time(usec_t until, uint32_t ttl, DnsRecordTTLState ttl_state) {
+        assert(ttl_state >= DNS_RECORD_TTL_STATE_80_PERCENT);
+        assert(ttl_state < _DNS_RECORD_TTL_STATE_MAX);
+
+        int percent = DNS_RECORD_TTL_STATE_TO_PERCENT(ttl_state);
+        assert(percent > 0);
+
+        return usec_sub_unsigned(until, (100 - percent) * ttl * USEC_PER_SEC / 100);
+}
```

Comment:
hmm, is this function used outside of this .c file? if so please do not export it. (same for all other funcs defind here). i.e. add static wherever you can.